core/src/main/java/com/graphhopper/util/PathSimplification.java

```diff
@@ -25,117 +25,252 @@
+    /**
+     * @see PathSimplification
+     */
+    private final List<Partition> partitions;
+    private final DouglasPeucker douglasPeucker;
 
-    public PathSimplification(PathWrapper pathWrapper, DouglasPeucker douglasPeucker, boolean enableInstructions) {
-        this.pointList = pathWrapper.getPoints();
-        listsToSimplify = new ArrayList<>();
-        if (enableInstructions)
-            listsToSimplify.add(pathWrapper.getInstructions());
+    // temporary variables used when traversing the different partitions
+    private final int numPartitions;
+    private final int[] currIntervalIndex;
+    private final int[] currIntervalStart;
+    private final int[] currIntervalEnd;
+    private final boolean[] partitionFinished;
+    // keep track of how many points were removed by Douglas-Peucker in the current and previous intervals
+    private final int[] removedPointsInCurrInterval;
+    private final int[] removedPointsInPrevIntervals;
 
-        this.pathDetails = pathWrapper.getPathDetails();
-        for (String name : pathDetails.keySet()) {
-            List<PathDetail> pathDetailList = pathDetails.get(name);
+    /**
+     * Convenience method used to obtain the partitions from a calculated path with details and instructions
+     */
+    public static PointList simplify(PathWrapper pathWrapper, DouglasPeucker douglasPeucker, boolean enableInstructions) {
+        final PointList pointList = pathWrapper.getPoints();
+        List<Partition> partitions = new ArrayList<>();
+        // todo: maybe this code can be simplified if path details and instructions would be merged, see #1121
+        if (enableInstructions) {
+            final InstructionList instructions = pathWrapper.getInstructions();
+            partitions.add(new Partition() {
+                @Override
+                public int size() {
+                    return instructions.size();
+                }
+
+                @Override
+                public int getIntervalLength(int index) {
+                    return instructions.get(index).getLength();
+                }
+
+                @Override
+                public void setInterval(int index, int start, int end) {
+                    Instruction instruction = instructions.get(index);
+                    if (instruction instanceof ViaInstruction || instruction instanceof FinishInstruction) {
+                        if (start != end) {
+                            throw new IllegalStateException("via- and finish-instructions are expected to have zero length");
+                        }
+                        // have to make sure that via instructions and finish instructions contain a single point
+                        // even though their 'instruction length' is zero.
+                        end++;
+                    }
+                    instruction.setPoints(pointList.shallowCopy(start, end, false));
+                }
+            });
+        }
```

Comment:
This is the fix compared to #1782. For via and finish instructions I increase the interval length, such that they contain a single point. Pretty ugly...

Reply:
But isn't this similar to how point count and edge count behave? E.g. a path with 1 edge contains 2 points and so 0 edge has 1 point? Hmmh, but yes. A JSON line of 1 point is not allowed.

Reply:
> But isn't this similar to how point count and edge count behave? E.g. a path with 1 edge contains 2 points and so 0 edge has 1 point? Hmmh, but yes. A JSON line of 1 point is not allowed.

Ah you mean instruction length = number of edges of the interval ? Yes this works, but like you say for a single point its a bit unclear.

Reply:
Yes. It is similar, so if you have a Via- or FinishInstruction it should still have a location (one point) but no extent or distance.